Сибирцев Иван

IvanSibirtsev opened the pull request:
@dmitgaranin

IvanSibirtsev committed:
project init and add algorithm from tdd

IvanSibirtsev committed:
delete unused classes

IvanSibirtsev committed:
add file read service

IvanSibirtsev committed:
add settings

IvanSibirtsev committed:
add words preparer

IvanSibirtsev committed:
add words to tag transformer

IvanSibirtsev committed:
add Drawable

IvanSibirtsev committed:
add Font, Size and Color services

IvanSibirtsev committed:
update ArhimedeanSpiralPointGenerator

IvanSibirtsev committed:
update CircularCloudLayouter.cs

IvanSibirtsev committed:
add tag to drawable tag transformer

IvanSibirtsev committed:
add image creator

IvanSibirtsev committed:
add image saver

IvanSibirtsev committed:
add visualizer

IvanSibirtsev committed:
add di container

IvanSibirtsev committed:
cleanup

IvanSibirtsev committed:
add settings provider

IvanSibirtsev committed:
add project to CLI and TagsCloudVisualization became library

IvanSibirtsev committed:
add CLI

IvanSibirtsev committed:
cleanup

dmitgaranin commented on TagsCloudCLI/Options.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using CommandLine;
```
**Дисклэймер:** если ты с чем-то не согласен или что-то хочешь обсудить - смело оставляй комменты здесь или приходи в лс телеграмма) но не забудь меня пингануть, что вопросы сформированы и я могу на них отвечать)

dmitgaranin commented on TagsCloudVisualization/TagsCloudModule.cs:
```diff
@@ -0,0 +1,49 @@
+        private readonly Settings settings;
+
+        public TagsCloudDrawerModule(Settings settings) =>
+            this.settings = settings ?? throw new ArgumentNullException(nameof(settings));
```
Эту проверку можно переложить на компилятор, если добавить в .csproj
```
<Nullable>enable</Nullable>
<WarningsAsErrors>Nullable</WarningsAsErrors>
```

dmitgaranin replied:
Предлагаю это использовать во всех проектах, а не только в указанном) полезная вещь

dmitgaranin commented on TagsCloudVisualization/TagsCloudModule.cs:
```diff
@@ -0,0 +1,49 @@
+            builder.RegisterInstance(new BoringWordsFilter(settings.BoringWords)).As<IWordsPreparer>();
+            builder.RegisterComposite<IWordsPreparer>((_, processors) => new WordsPreparer(processors));
+            builder.RegisterInstance(new TagColorService(settings.TagColor)).As<ITagColorService>();
+            builder.RegisterType<WordsToTagTransformer.WordsToTagTransformer>().As<IWordsToTagTransformer>();
```
В этом месте и ниже пришлось прописывать namespace, чтобы достучаться до нужного класса. Выглядит не очень хорошо. Можешь как-нибудь красиво поправить?

dmitgaranin commented on TagsCloudVisualization/TagsCloudModule.cs:
```diff
@@ -0,0 +1,49 @@
+                    ctx.Resolve<IEnumerable<IWordsReader>>()))
+                .As<IFileReadService>();
+            builder.RegisterType<WordsToLowerPrepare>().As<IWordsPreparer>();
+            builder.RegisterInstance(new BoringWordsFilter(settings.BoringWords)).As<IWordsPreparer>();
```
Кажется, подход, когда регистрация происходит на основе внешней зависимости, не очень хорош. Возможно, есть способ сделать так, чтобы и настройки лежали внутри DI. Вдруг это улучшит читаемость)

dmitgaranin commented on TagsCloudVisualization/TagsCloudModule.cs:
```diff
@@ -0,0 +1,49 @@
+            builder.Register(_ => new TagFontService(settings.MaxFontSize, settings.FontFamilyName))
+                .As<ITagFontService>();
+            builder.RegisterType<TagSizeService>().As<ITagSizeService>();
+            builder.RegisterType<TagToDrawableTagTransformer>().AsSelf();
```
Можешь рассказать, почему часть объектов ты регистрировал через интерфейс, а часть - нет? Кажется, упускаются точки расширения

dmitgaranin commented on TagsCloudVisualization/TagsCloudModule.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace TagsCloudVisualization
+{
+    public class TagsCloudDrawerModule : Module
```
Почему-то класс и файл имеют разные имена

dmitgaranin commented on TagsCloudVisualization/WordsProvider/FileReadService.cs:
```diff
@@ -0,0 +1,34 @@
+
+        public IEnumerable<string> Read()
+        {
+            var extension = Path.GetExtension(path);
```
Думаю, эта операция будет вызываться неоправданно много, если запускать чтение многократно

dmitgaranin commented on TagsCloudVisualization/WordsProvider/FileReader/IWordsReader.cs:
```diff
@@ -0,0 +1,8 @@
+{
+    public interface IWordsReader
+    {
+        bool CanRead(string extension);
```
Возможно, больше подойдет название `IsSupportedFileExtension`

dmitgaranin commented on TagsCloudVisualization/WordsProvider/FileReader/IWordsReader.cs:
```diff
@@ -0,0 +1,8 @@
+    public interface IWordsReader
+    {
+        bool CanRead(string extension);
+        string Read(string filename);
```
А здесь, мб, `GetFileContent`, т.к. `read` говорит больше о процессе, чем о результате. Например, не очень очевидно, что возвращается содержимое всего файла, а не какая-то его часть или вообще stream

dmitgaranin replied:
Также полагаю, что здесь передается не имя файла, а путь до файла

dmitgaranin commented on TagsCloudVisualization/WordsProvider/FileReadService.cs:
```diff
@@ -0,0 +1,34 @@
+            var reader = readers.FirstOrDefault(reader => reader.CanRead(extension));
+            if (reader == null)
+                throw new ArgumentException($"Unsupported file extension: {extension}");
+            var text = reader.Read(path);
```
А не может быть ситуации, что между созданием `FileReadService` и чтением файл будет удален?

dmitgaranin commented on TagsCloudVisualization/WordsProvider/FileReadService.cs:
```diff
@@ -0,0 +1,34 @@
+            if (reader == null)
+                throw new ArgumentException($"Unsupported file extension: {extension}");
+            var text = reader.Read(path);
+            return Regex.Split(text, WordsSplitPattern);
```
Выглядит неоптимизированным на многократном вызове чтения

dmitgaranin commented on TagsCloudVisualization/WordsPrepare/IWordsPreparer.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using System.Collections.Generic;
+
+namespace TagsCloudVisualization.WordsPrepare
```
А namespace может содержать глаголы?

dmitgaranin commented on TagsCloudVisualization/WordsPrepare/Preparers/BoringWordsFilter.cs:
```diff
@@ -0,0 +1,17 @@
+        public IEnumerable<string> Prepare(IEnumerable<string> words) =>
+            words.Where(word => !IsBoring(word) && word != "");
+
+        private bool IsBoring(string word) => boringWords.Contains(word);
```
Полагаю, что никто не давал гарантий, что скучные слова будут в нижнем регистре

dmitgaranin commented on TagsCloudVisualization/WordsPrepare/Preparers/BoringWordsFilter.cs:
```diff
@@ -0,0 +1,17 @@
+        public BoringWordsFilter(IEnumerable<string> boringWords) => this.boringWords = boringWords.ToHashSet();
+
+        public IEnumerable<string> Prepare(IEnumerable<string> words) =>
+            words.Where(word => !IsBoring(word) && word != "");
```
Давай заменим `""` на `string.Empty` . Или наверно в твоем случае больше подойдет вообще `string.IsNullOrWhiteSpace`

dmitgaranin replied:
Так же, возможно стоит сперва проверить слово на пустоту, а уже потом на скучность)

dmitgaranin commented on TagsCloudVisualization/WordsPrepare/WordsPreparer.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace TagsCloudVisualization.WordsPrepare
+{
+    public class WordsPreparer : IWordsPreparer
```
Полагаю, этот класс не является `IWordsPreparer`, т.к. он объединяет `IWordsPreparer`. Возможно, есть чуть лаконичнее способ зачейнить `IWordsPreparer`

dmitgaranin commented on TagsCloudVisualization/Visualizer.cs:
```diff
@@ -0,0 +1,41 @@
+            var words = fileReadService.Read();
+            var preparedWords = wordsPreparer.Prepare(words);
+            var tags = wordsToTagTransformer.Transform(preparedWords);
+            var drawableTags = tagToDrawableTagTransformer.Transform(tags.ToList());
```
А давай либо сделаем по-другому, либо поменяем тип параметра у `tagToDrawableTagTransformer.Transform` ?)

dmitgaranin commented on TagsCloudVisualization/TagToDrawableTagTransformer.cs:
```diff
@@ -0,0 +1,44 @@
+            var min = tags.Min(tag => tag.Count);
+            var max = tags.Max(tag => tag.Count);
+            var drawables = new List<IDrawable>();
+            foreach (var tag in tags.OrderByDescending(x => x.Count))
```
В этом методе принимаемый `IEnumerable` итерируется три раза. Можешь рассказать, какие есть плюсы и минусы у этого?

dmitgaranin commented on TagsCloudVisualization/TagToDrawableTagTransformer.cs:
```diff
@@ -0,0 +1,44 @@
+            var drawables = new List<IDrawable>();
+            foreach (var tag in tags.OrderByDescending(x => x.Count))
+            {
+                var font = tagFontService.GetFont(tag, min, max);
```
А давай вынесем создание `DrawableTag` в метод? Думаю, это сильно повысит читаемость

dmitgaranin commented on TagsCloudVisualization/ImageCreator/ImageCreator.cs:
```diff
@@ -0,0 +1,24 @@
+
+            graphics.TranslateTransform(bitmap.Width / 2f, bitmap.Height / 2f);
+
+            foreach (var drawableTag in layout)
```
Множественная итерация по `IEnumerable`

dmitgaranin commented on TagsCloudVisualization/TagsCloudVisualization.csproj:
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net5.0</TargetFramework>
```
В твоем коде используются штуки, специфичные только для платформы windows

dmitgaranin commented on TagsCloudVisualization/Saver/ImageSaver.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public void Save(Image image)
+        {
+            var pathToDirectory = Path.Combine(Directory.GetCurrentDirectory(), directory);
```
Подозреваю, что ответственность за предоставление пути до директории лежит на потребители приложения, а не на самом приложении. Как минимум будет очень неочевидно, куда будет сохранен результат

dmitgaranin commented on TagsCloudVisualization/Visualizer.cs:
```diff
@@ -0,0 +1,41 @@
+
+        public void Visualize()
+        {
+            var words = fileReadService.Read();
```
(на твое усмотрение)
Т.к. все методы ниже принимают результаты предыдущей функции в параметрах, то, может, их получится зачейнить

dmitgaranin commented on TagsCloudVisualization/Visualizer.cs:
```diff
@@ -0,0 +1,41 @@
+            this.imageSaver = imageSaver;
+        }
+
+        public void Visualize()
```
Давай сменим название? От текущего не ожидаешь, что он 
- не выдает результат
- сохраняет результат "куда-то"

Либо разбить на визуализатор и сохранятор

dmitgaranin commented on TagsCloudVisualization/Saver/ImageSaver.cs:
```diff
@@ -0,0 +1,28 @@
+        {
+            var pathToDirectory = Path.Combine(Directory.GetCurrentDirectory(), directory);
+            var path = Path.Combine(pathToDirectory, imageName);
+            image.Save(path, ImageFormat.Bmp);
```
> В качестве результата программа должна генерировать png-файл.

dmitgaranin commented on TagsCloudCLI/Program.cs:
```diff
@@ -0,0 +1,38 @@
+    { 
+        static void Main(string[] args)
+        {
+            HandleErrors(() =>
```
Выглядит усложнением. Думаю, можно было не выносить try-catch в метод

dmitgaranin replied:
Можешь рассказать, почему решил оставить?)

dmitgaranin commented on TagsCloudCLI/SettingProvider.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace TagsCloudCLI
+{
+    public class SettingProvider
```
Он точно должен быть public?

dmitgaranin commented on TagsCloudVisualization/Extensions/IDrawableExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace TagsCloudVisualization.Extensions
+{
+    public static class DrawableExtensions
```
Он точно должен быть паблик?

IvanSibirtsev committed:
fix TagToDrawableTagTransformer:
1. change input arguments (IEnumerable -> List)
2. define min and max in one iteration
3. create interface ITagToDrawableTransformer
4. fix Visualizer and di-container to work with it

IvanSibirtsev committed:
Image saver and Options change bmp to png

IvanSibirtsev committed:
Options and SettingProvider change public to internal

IvanSibirtsev committed:
fix FileReadService:
1. optimize regex and getExtension
2. check file exists
3. rename methods

IvanSibirtsev committed:
add settings to di-container

IvanSibirtsev committed:
rename namespaces

IvanSibirtsev committed:
rename to TagsCloudModule

IvanSibirtsev committed:
fix WordsPreparer:
1. rename WordsPreparer to WordsPreprocessor
2. divide IWordsPreparer to IWordsPreprocessor, IWordsPreparer and IWordsFilter
3. redone check emptiness of string
4. fix Visualizer and di-container to work with it

IvanSibirtsev committed:
fix double enumeration
1. add Drawable container builder and drawable container
2. dix Visualizer and di-container to work with it

IvanSibirtsev committed:
fix settings:
1. add short names
2. chang help text
3. make Directory and File with words as required arguments

IvanSibirtsev committed:
fix image saver
1. divide visualizer to visualize and save
2. add different image format support
3. fix Visualizer and di-container to work with it

IvanSibirtsev committed:
add classes to specific settings

IvanSibirtsev committed:
add nullable check to .csproj

IvanSibirtsev committed:
fix . in extension name

IvanSibirtsev committed:
make project only for windows

IvanSibirtsev committed:
add new readers

IvanSibirtsev committed:
add tests

IvanSibirtsev committed:
add ability to get boring words from file

dmitgaranin commented on TagsCloudCLI/ImageSaver.cs:
```diff
@@ -0,0 +1,42 @@
+
+        public void Save(Image image)
+        {
+            var pathToDirectory = Path.Combine(Directory.GetCurrentDirectory(), directory);
```
Выше ты проверяешь, что переданный путь до директории валиден, а затем ты его откладываешь от текущей папки? Тогда зачем ты требуешь указывать путь до директории в параметрах?

dmitgaranin commented on TagsCloudCLI/ImageSaver.cs:
```diff
@@ -0,0 +1,42 @@
+            format = DefineFormat();
+        }
+
+        private ImageFormat DefineFormat()
```
Если в нее передавать имя, то можно:
- сделать статиком
- вынести в другой файл (потенциально)

dmitgaranin commented on TagsCloudCLI/Program.cs:
```diff
@@ -0,0 +1,40 @@
+                if (result.Errors.Any())
+                    return;
+                var settings = new SettingProvider().GetSettings(result.Value);
+                var imageSaver = new ImageSaver(settings.Saver.Directory, settings.Saver.ImageName);
```
А если добавится еще один клиент (например, GUI), который будет сохранять картинку?)

dmitgaranin commented on TagsCloudCLI/SettingProvider.cs:
```diff
@@ -0,0 +1,29 @@
+        
+        
+        private static IEnumerable<string> GetBoringWordsFromFile(string filename) =>
+            !File.Exists(filename) ? Array.Empty<string>() : File.ReadLines(filename);
```
А если будет передан невалидный путь до скучный слов? Такая логика не поможет пользователю об этом узнать

dmitgaranin commented on TagsCloudVisualization/DrawableContainers/IDrawableContainer.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface IDrawableContainer
+    {
+        IEnumerable<IDrawable> GetDrawables();
```
Не очень понятно, что именно возвращает метод

dmitgaranin commented on TagsCloudVisualization/ImageCreators/ImageCreator.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace TagsCloudVisualization.ImageCreators
+{
+    public class ImageCreator : IImageCreator
```
Этот и другие объекты лучше спрятать под internal

dmitgaranin commented on TagsCloudVisualization/TagToDrawableTransformer/TagToDrawableTagTransformer.cs:
```diff
@@ -0,0 +1,59 @@
+            this.layouter = layouter;
+        }
+
+        public IEnumerable<IDrawable> Transform(List<Tag> tags)
```
Программа не работает, т.к. реализуется неправильный интерфейс
`public IEnumerable<IDrawable> Transform(IEnumerable<Tag> tags);`

dmitgaranin commented on TagsCloudVisualizationTests/FunctionalTest/TagsCloudTest.cs:
```diff
@@ -0,0 +1,39 @@
+    public class TagsCloudTest
+    {
+        [TestCase("txt", TestName = "txt")]
+        public void Should_ReadWords_From(string extension)
```
Тест падает 
![image](https://user-images.githubusercontent.com/8987952/145986168-fe4c7575-8926-4086-b552-bd5a68bfd395.png)

dmitgaranin commented on TagsCloudVisualizationTests/FunctionalTest/TagsCloudTest.cs:
```diff
@@ -0,0 +1,39 @@
+            var builder = new ContainerBuilder();
+            builder.RegisterModule(new TagsCloudModule(settings));
+            var container = builder.Build();
+            var imageSaver = new ImageSaver(settings.Saver.Directory, settings.Saver.ImageName);
```
Предлагаю не добавлять в тесты зависимость от консольного клиента только ради ImageSaver. Полагаю, этот класс должен лежать в другом месте

dmitgaranin commented on TagsCloudVisualizationTests/FunctionalTest/TagsCloudTest.cs:
```diff
@@ -0,0 +1,39 @@
+            var visualizer = container.Resolve<Visualizer>();
+            var image = visualizer.Visualize();
+            
+            imageSaver.Save(image);
```
А как гарантируется, что сохранилось что-то правильное? Насколько я могу судить по тесту, он ничего не проверяет, а только запускает визуализатор с сохранялкой

dmitgaranin commented on TagsCloudVisualizationTests/PreprocessTest/FilterTest.cs:
```diff
@@ -0,0 +1,20 @@
+
+            var processed = filter.Filter(words);
+
+            processed.Should().ContainInOrder("text", "text2");
```
Эта проверка не гарантирует, что в результате нет лишних слов. Она проверяет, что результат _содержит_ эти два слова. Лучше переделать на что-то подобное 
```
processed.Should().BeEquivalentTo(new []{ "text", "text2"}, options => options.WithStrictOrdering());
```

dmitgaranin replied:
и давай еще добавим кейс, когда фильтрация игнорируют регистр

dmitgaranin commented on TagsCloudVisualizationTests/PreprocessTest/PrepareTest.cs:
```diff
@@ -0,0 +1,20 @@
+
+            var processed = preparer.Prepare(words);
+
+            processed.Should().ContainInOrder("text", "a", "b", "text2");
```
Аналогично комментарию выше

dmitgaranin commented on TagsCloudVisualizationTests/PreprocessTest/PreprocessTest.cs:
```diff
@@ -0,0 +1,24 @@
+
+            var processed = preprocessor.Preprocess(words);
+
+            processed.Should().ContainInOrder("text", "text2");
```
И тут

dmitgaranin commented on TagsCloudVisualizationTests/WordsToTagTransformersTest/WordsToTagTransforemerTest.cs:
```diff
@@ -0,0 +1,34 @@
+            var tags = transformer.Transform(words).ToList();
+
+            tags.Should().HaveCount(3);
+            tags
```
Давай лучше сравним с ожидаемым результатом, т.к. этот тест не гарантирует, что ответ содержит исходные слова - там может оказаться что-то другое

IvanSibirtsev committed:
public -> internal

IvanSibirtsev committed:
add ImageSaver and FunctionalTest .project

IvanSibirtsev committed:
delete TagToDrawableTagTransformer

IvanSibirtsev committed:
fix Options.cs and SettingProvider

IvanSibirtsev committed:
fix Program main

IvanSibirtsev committed:
add default txt

IvanSibirtsev committed:
fix PreprocessTest, FilterTest and PrepareTest

IvanSibirtsev committed:
add dependencies

IvanSibirtsev committed:
add Functional test

IvanSibirtsev committed:
fix WordsToTagTransformerTest

IvanSibirtsev committed:
fix nullable

IvanSibirtsev committed:
fix dependencies

IvanSibirtsev committed:
fix main

IvanSibirtsev committed:
small fixes

dmitgaranin commented on TagsCloudVisualization/WordsPreprocessors/Preparers/WordsToLowerPrepare.cs:
```diff
@@ -0,0 +1,11 @@
+    public class WordsToLowerPrepare : IWordsPreparer
+    {
+        public IEnumerable<string> Prepare(IEnumerable<string> words) =>
+            words.Select(word => word.ToLower());
```
Думаю, метод приведения к нижнему регистру должен повторять логику и в скучных словах. Или, что может быть лучше, в скучных словах создать хэш-мапу через `new HashSet<string>(set, StringComparer.OrdinalIgnoreCase)`

dmitgaranin commented on TagsCloudVisualization/Extensions/IDrawableExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace TagsCloudVisualization.Extensions
+{
+    internal static class DrawableExtensions
```
Не используется

dmitgaranin commented on CloudImageSaver/CloudImageSaver.csproj:
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net5.0-windows</TargetFramework>
```
Ему не нужна зависимость от win
